### PR TITLE
Add development verify scripts and refactor Dockerfile

### PR DESCRIPTION
**Pull request by zryfish**
### What type of PR is this?

/kind cleanup

### What this PR does / why we need it:

This PR is part of our continued work #3911, it mainly does the following things.

* rewrite Dockerfile with docker multiple-stage build to keep it simple and flexible
* add verification scripts to avoid obvious mistakes before submitting, most of verify scripts are borrowed from `kubernetes/kubernetes` repo.

Also, it's easier to migrate our CI workflow to Prow once this PR is merged. 

1. `make verify` can be used to verify PR, make sure no obvious typos, mistakes.
2. `make container` can be used to  build `ks-apiserver` `ks-controller-manager` docker images without pushing to remote, while `make container-push` will build and push.
3. `make container-cross` and `make container-cross-push` are same with `make container`, only `make container-cross` will build images for platform `linux/amd64` and `linux/arm64`. `make container-cross-push` will build and push multiple platform images.

### Which issue(s) this PR fixes:

Fixes #

### Special notes for reviewers:
```

```

### Does this PR introduced a user-facing change?

```release-note
None
```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

File: build/ks-apiserver/Dockerfile [hunk 1]
```diff
@@ -1,23 +1,49 @@
...
+FROM golang:1.16.3 as build_context
+
+ENV OUTDIR=/out
+RUN mkdir -p ${OUTDIR}/usr/local/bin/
+
+WORKDIR /workspace
+ADD . /workspace/
+
+RUN make ks-apiserver
+RUN pwd && ls -alh
```
Comment (zryfish) [inline build/ks-apiserver/Dockerfile:34]: This line is unnecessary,  will be removed later.

**Comment by yuswift** [comment]
> ### What type of PR is this?
> /kind cleanup
> 
> ### What this PR does / why we need it:
> This PR is part of our continued work #3911, it mainly does the following things.
> 
> * rewrite Dockerfile with docker multiple-stage build to keep it simple and flexible
> * add verification scripts to avoid obvious mistakes before submitting, most of verify scripts are borrowed from `kubernetes/kubernetes` repo.
> 
> Also, it's easier to migrate our CI workflow to Prow once this PR is merged.
> 
> 1. `make verify` can be used to verify PR, make sure no obvious typos, mistakes.
> 2. `make container` can be used to  build `ks-apiserver` `ks-controller-manager` docker images without pushing to remote, while `make container-push` will build and push.
> 3. `make container-cross` and `make container-cross-push` are same with `make container`, only `make container-cross` will build images for platform `linux/amd64` and `linux/arm64`.
> 
> ### Which issue(s) this PR fixes:
> Fixes #
> 
> ### Special notes for reviewers:
> ```
> 
> ```
> 
> ### Does this PR introduced a user-facing change?
> ```
> None
> ```
> 
> ### Additional documentation, usage docs, etc.:
> ```
> 
> ```

4.`make container-cross-push` will build and push.

**Comment by yuswift** [comment]
@zryfish 
what about executing `make container`  here https://github.com/kubesphere/kubesphere/blob/d1fdc7c9e072dfc39ae7dabb5c30963cc9e6d5e4/.github/workflows/build.yml#L46
somebody may change the Dockerfile, we need to verify that. In the meanwhile, this may make the building time longer. 
We also need to execute `make verify` in the workflow or prow.

**Comment by zryfish** [comment]
> @zryfish
> what about executing `make container` here
> 
> https://github.com/kubesphere/kubesphere/blob/d1fdc7c9e072dfc39ae7dabb5c30963cc9e6d5e4/.github/workflows/build.yml#L46
> 
> 
> somebody may change the Dockerfile, we need to verify that. In the meanwhile, this may make the building time longer.
> We also need to execute `make verify` in the workflow or prow.

`make verify` will give tons of errors now, I would like to keep the Github workflow the same, and enable `make verify` when we finished migrating to Prow, it will give us time to fix `verify` errors.

`make container` only build image, if you inspect the Makefile, you will find `make container` doesn't execute `make test`. So `make all` is necessary here, only `make all` passed we can `make container`.

**Comment by yuswift** [comment]
sorry for the mistake... 
I forgot that jeff was the approver.

**Pull request merged**